syscall.1
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.29)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "syscall 1"
.TH syscall 1 "2017-07-03" "1.0" ""
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
syscall \- send system calls from your shell
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&  syscall [\-<n>] name [args...] [, name [args...]]...
\&
\&  syscall [\-h|\-\-help]
\&
\&  syscall [\-v|\-\-version]
.Ve
.PP
To replicate the following C fragment:
.PP
.Vb 3
\&  int fd = open("/my/file", O_WRONLY, 0755);
\&  write(fd, "hello", strlen("hello"));
\&  close(fd);
.Ve
.PP
you would use syscall like this:
.PP
.Vb 1
\&  syscall open /my/file 1 0755 , write \e$0 hello \e#hello , close \e$0
.Ve
.PP
To print the return code of a system call, use \f(CW\*(C`echo\*(C'\fR:
.PP
.Vb 1
\&  syscall open /dev/random 0 , echo \e$0
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Execute a list of raw system calls. All the system calls listed in your
system's unistd.h are supported, with up to 5 arguments. A maximum of 20
calls can be executed per invocation, each separated by a comma.
.PP
Arguments starting by a \f(CW\*(C`#\*(C'\fR symbol are used to give a string length. For
instance, \f(CW\*(C`#hello\*(C'\fR would be evaluated as 5.
.PP
Arguments starting by a \f(CW\*(C`$\*(C'\fR followed by a number from 0 to 19 refer to a
previous system call return code. For instance, \f(CW$0\fR refers to to the return
code of the first system call executed. To display those values, use the
\&\f(CW\*(C`echo\*(C'\fR built-in command.
.PP
The \f(CW\*(C`echo\*(C'\fR command can be used like any other system call to easily
display \f(CW\*(C`$\*(C'\fR or \f(CW\*(C`#\*(C'\fR values, or any string or number.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-<n\fR>" 4
.IX Item "-<n>"
.Vb 2
\&  # print "foo" 10 times
\&  syscall \-10 write 1 "foo\en" 4
.Ve
.Sp
Execute the given commands \fIn\fR times, where \fIn\fR is an integer between 0
and \f(CW\*(C`INT_MAX\*(C'\fR.
.IP "\fB\-h \-\-help\fR" 4
.IX Item "-h --help"
.PD 0
.IP "\fB\-v \-\-version\fR" 4
.IX Item "-v --version"
.PD
.SH "EXIT STATUS"
.IX Header "EXIT STATUS"
\&\f(CW0\fR if all syscalls were successful, \f(CW1\fR on error.
.PP
Note that if any system returns \-1, the program will exit immediately after
printing the associated error message.
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright 2017 Olivier Duclos.
.PP
This program is distributed under the \s-1ISC\s0 license.
.PP
Heavily inspired by the syscall command from \s-1AIX.\s0
.SH "SEE ALSO"
.IX Header "SEE ALSO"
Linux Programmer's manual : \fIsyscall\fR\|(2)
.PP
This project's homepage: <https://github.com/oliwer/syscall>


syscall.c
/*
  syscall - Launch system calls from your shell
  Copyright (c) 2017, Olivier Duclos

  Permission to use, copy, modify, and/or distribute this software for any
  purpose with or without fee is hereby granted, provided that the above
  copyright notice and this permission notice appear in all copies.

  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
  REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
  AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
  LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
  OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
  PERFORMANCE OF THIS SOFTWARE.
*/

#include <err.h>
#include <errno.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "systab.h"

#define LEN(x)      (sizeof(x) / sizeof(*x))
#define streq(a,b)  (strcmp(a,b) == 0)

#define CMD_MAX   20  /* maximum number of commands per invocation */

static int ret_values[CMD_MAX];

void usage()
{
  puts("usage: syscall [-<n>] name [args...] [, name [args...]]...");
}

int scomp(const void *m1, const void *m2)
{
  Syscall *sys1 = (Syscall *)m1;
  Syscall *sys2 = (Syscall *)m2;
  return strcmp(sys1->name, sys2->name);
}

long str2syscall(const char *name)
{
  Syscall key, *res;

  key.name = (char *)name;

  res = bsearch(&key, systab, LEN(systab), sizeof key, scomp);
  if (res == NULL) {
    errx(1, "unknown system call: %s", name);
  }

  return (long)res->code;
}

/* Quick and dirty way to unescape \n at the end of strings */
void unescape_nl(char *str) {
  size_t end = strlen(str) - 1;

  while (str[end] == 'n' && str[end-1] == '\\') {
    str[end-1] = '\n';
    str[end] = '\0';
    end -= 2;
  }
}

unsigned long parse_arg(const char *syscall_name, char *arg)
{
  unsigned long num;
  char *endp = NULL;

  /* #hello - length of a string */
  if (arg[0] == '#') {
    return (unsigned long)strlen(arg + 1);
  }

  /* $0 - return value of a previous syscall */
  if (arg[0] == '$') {
    num = strtoul(arg + 1, &endp, 10);
    if (num < CMD_MAX - 1) {
      return (unsigned long)ret_values[num];
    }
  }

  /* Try to parse it as a number */
  endp = NULL;
  num = strtoul(arg, &endp, 0);
  if (errno == ERANGE) {
    errx(1, "%s: argument '%s' is out of range",
      syscall_name, arg);
  }
  if (endp && *endp == '\0') {
    /* strtoul succeeded */
    return num;
  }

  /* assume it is a string */
  /* unescape any \n at the end of the string */
  unescape_nl(arg);
  return (unsigned long)arg;
}

void echo(int argc, char **argv) {
  int i;

  for (i = 1; i < argc; i++) {
    if (argv[i][0] == '$' || argv[i][0] == '#') {
      printf("%ld\n", parse_arg(__func__, argv[i]));
    } else {
      printf("%s\n", argv[i]);
    }
  }
}

#define ARG(n)    parse_arg(syscall_name, cmd[n])

void parse_syscall(int cmd_no, char **cmd, int cmd_len)
{
  long syscall_num, ret = -1;
  char *syscall_name = cmd[0];

  /* Special case */
  if (streq(syscall_name, "echo")) {
    echo(cmd_len, cmd);
    ret_values[cmd_no] = 0; /* echo never fails */
    return;
  }

  syscall_num = str2syscall(syscall_name);
  if (syscall_num == -1) {
    errx(1, "unknown system call '%s'", syscall_name);
  }

  switch (cmd_len) {
  case 0:
    ret = syscall(syscall_num);
    break;
  case 1:
    ret = syscall(syscall_num, ARG(1));
    break;
  case 2:
    ret = syscall(syscall_num, ARG(1), ARG(2));
    break;
  case 3:
    ret = syscall(syscall_num, ARG(1), ARG(2), ARG(3));
    break;
  case 4:
    ret = syscall(syscall_num, ARG(1), ARG(2), ARG(3), ARG(4));
    break;
  case 5:
    ret = syscall(syscall_num, ARG(1), ARG(2), ARG(3), ARG(4), ARG(5));
    break;
  default:
    errx(1, "too many arguments for %s", syscall_name);
  }

  if (ret == -1) {
    errx(1, "%s failed: %s", syscall_name, strerror(errno));
  }

  ret_values[cmd_no] = (int)ret;
}

void split_cmdline(int argc, char **argv)
{
  int cmd_no = 0, cmd_len = 0, i;

  for (i = 0; i < argc; i++) {
    //printf("dbg: %s\n", argv[i]);
    if (cmd_no == CMD_MAX) {
      errx(1, "too many command (%d > %d)", cmd_no, CMD_MAX);
    }
    if (streq(argv[i], ",") || i == argc-1) {
      //printf("parse: len=%d  no=%d  start=%s\n", cmd_len, cmd_no, (argv + (i - cmd_len))[0]);
      parse_syscall(cmd_no, argv + (i - cmd_len), cmd_len);
      cmd_no++;
      cmd_len = 0;
    } else {
      cmd_len++;
    }
  }
}

/* for debugging */
void dump_ret_values(void) {
  int i;

  for (i = 0; i < CMD_MAX; i++) {
    printf("%0d  %d\n", i, ret_values[i]);
  }
}

int main(int argc, char **argv)
{
  int repeat = 1, skip = 1;

  if (argc <= 1) {
    usage();
    return 0;
  }

  /* arguments */
  if (argv[1][0] == '-') {
    /* help */
    if (streq(argv[1], "-h") || streq(argv[1], "--help")) {
      usage();
      return 0;
    }

    /* version */
    if (streq(argv[1], "-v") || streq(argv[1], "--version")) {
      puts("syscall version "VERSION);
      return 0;
    }

    /* Handle -n option */
    repeat = atoi(argv[1] + 1);
    if (repeat < 1) {
      errx(1, "option -<n> must be between 1 and %d", INT_MAX);
    }
    if (argc <= 3) {
      usage();
      return 1;
    }
    skip++;
  }

  memset(ret_values, -1, sizeof ret_values);

  /* Prepare the syscalls table for searching */
  qsort(systab, LEN(systab), sizeof(Syscall), scomp);

  while (repeat--) {
    split_cmdline(argc - skip, argv + skip);
  }

  return 0;
}
